slowfast/models/build.py

```diff
@@ -26,6 +27,95 @@
 """
 
 
+def get_fsdp_params(cfg) -> dict:
+    """
+    Return FSDP params dict (Assumes usage of fairscale FSDP)
+
+    Args:
+        cfg: config containing FSDP wrapping params
+
+    Return:
+        fsdp_params: FSDP params dict
+    """
+    fsdp_params = {}
+    try:
+        fsdp_params['reshard_after_forward'] = cfg.FSDP.RESHARD_AFTER_FW
+        fsdp_params["mixed_precision"] = cfg.TRAIN.MIXED_PRECISION
+    except AttributeError as e:
+        logger.exception(f"Configuration error: {e}")
+        raise e
+    if cfg.FSDP.AUTO_WRAP:
+        fsdp_params['auto_wrap_policy'] = functools.partial(
+            default_auto_wrap_policy,
+            min_num_params=int(cfg.FSDP.MIN_PARAMS_TO_WRAP)
+        )
+
+    return fsdp_params
+
+
+def fsdp_model(model: torch.nn.Module, cfg: dict, cur_device: torch.device):
+    """
+    Wraps a model with FSDP.
+
+    Args:
+        model: unwrapped torch model
+        cfg: config containing FSDP config params
+        cur_device: The device to which the model should be transfered
+
+    Return:
+        model: FSDP wrapped model
+    """
+    assert not (cfg.FSDP.AUTO_WRAP and cfg.FSDP.NESTED_WRAP), "FSDP mode: select either AUTO_WRAP or NESTED_WRAP"
+    fsdp_params = get_fsdp_params(cfg)
+
+    if cfg.FSDP.AUTO_WRAP:
+        with enable_wrap(wrapper_cls=FSDP, **fsdp_params):
+            model = auto_wrap(model)
```

Comment:
You can't wrap a model after it has been created, otherwise nothing inside it will get wrapped. The model has to be created in the `enable_wrap` context. Can you see submodules wrapped in FSDP when you print the model currently?

Reply:
To my understanding, `auto_wrap` actually calls the `wrap` function recursively on all children modules if the policy is satisfied and the `enable_wrap` context exists. Reference to the code [here](https://github.com/facebookresearch/fairscale/blob/c1dada48b2e5a91394edb88980dac45f66e14e1e/fairscale/nn/wrap/auto_wrap.py#L321). Attached is a screenshot of an auto wrapped model.

<img width="1054" alt="Screen Shot 2022-08-02 at 8 30 52 AM" src="https://user-images.githubusercontent.com/8292193/182413658-39c8cc8a-27c0-4acd-bc09-9676ccec35db.png">

Reply:
Oh that's great! Yeah, sorry I mixed up nested wrapping and auto wrapping. Thanks for double checking!